File: tdd/scala/scalatest/test/MoneyTest.scala
import org.scalatest.FlatSpec
import org.scalatest.matchers.ShouldMatchers
import org.scalatest.matchers.MatchResult
import org.scalatest.matchers.Matcher
import org.scalatest.matchers.BeMatcher
import org.mockito.Mockito._
import lib._

package test {

class MoneyTest extends FlatSpec with ShouldMatchers {
    val oneDollar = new lib.Money(1, 'dollar)
    val twoFrancs = new lib.Money(2, 'franc)

    it should "be able to add dollars to dollars" in {
        (oneDollar + oneDollar) should be === (new Money(2, 'dollar))
        (oneDollar + oneDollar) should be <  (new Money(3, 'dollar))
        (oneDollar + oneDollar) should be <= (new Money(2, 'dollar))
    }

    it should "be able to subtract dollars to dollars" in {
        (new Money(2, 'dollar) - oneDollar) should be === (new Money(1, 'dollar))
    }

    it should "be unable to compare dollars and francs" in {
        oneDollar should not be === (twoFrancs)
        oneDollar should not be >  (twoFrancs)
        oneDollar should not be >= (twoFrancs)
        oneDollar should not be <  (twoFrancs)
        oneDollar should not be <= (twoFrancs)
    }

    it should "be unable to add dolars and francs" in {
        evaluating { oneDollar + twoFrancs } should produce [InvalidMoneyOperation]
    }

    it should "be unable to subtract dolars and francs" in {
        evaluating { oneDollar - twoFrancs } should produce [InvalidMoneyOperation]
    }

    it should "be able to apply the money on a bank" in {
        val bank = mock(classOf[Bank])
        oneDollar apply_on bank
        verify(bank) apply oneDollar
    }

    it should "be expensive if greater than 20,000" in {
        val thousand_bucks = new Money(30000, 'dollar)
        thousand_bucks should be (expensive)
    }

    it should "identify the currency of the money" in {
        twoFrancs should haveCurrency ('franc)
    }

    def expensive = new BeExpensiveMatcher()
    class BeExpensiveMatcher extends BeMatcher[Money] {
        def apply(money: Money) = MatchResult(
            money.value > 20000,
            money + " is not expensive",
            money + " is expensive"
        )
    }

    def haveCurrency(name: Symbol) = new HaveCurrencyMatcher(name)
    class HaveCurrencyMatcher(name: Symbol) extends Matcher[Money] {
        def apply(money: Money) = MatchResult(
            money.name == name,
            "Expected money to be of currency " + name,
            "Expected money to not be of currency " + name
        )
    }
}

}

File: tdd/java/junit/test/MoneyTest.java
package test;

import org.junit.*;
import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.*;
import lib.*;
import static org.mockito.Mockito.*;

public class MoneyTest {
    private Money oneDollar;
    private Money twoFrancs;

    @Before public void setUp() {
        oneDollar = new Money(1, "dollar");
        twoFrancs = new Money(2, "franc");
    }

    @Test public void addDollarsToDollars() throws InvalidMoneyOperation {
        assertThat(oneDollar.add(oneDollar), is(new Money(2, "dollar")));
        assertTrue((oneDollar.add(oneDollar)).lt(new Money(3, "dollar")));
        assertTrue((oneDollar.add(oneDollar)).lte(new Money(2, "dollar")));
    }

    @Test public void subtractDollarsToDollars() throws InvalidMoneyOperation {
        assertThat(new Money(2, "dollar").sub(oneDollar), is(new Money(1, "dollar")));
    }

    @Test public void cantCompareDollarsAndFrancs() {
        assertThat(oneDollar, is(not(twoFrancs)));
        assertFalse(oneDollar.gt(twoFrancs));
        assertFalse(oneDollar.gte(twoFrancs));
        assertFalse(oneDollar.lt(twoFrancs));
        assertFalse(oneDollar.lte(twoFrancs));
    }

    @Test(expected=InvalidMoneyOperation.class) public void cantAddDollarsAndFrancs() throws InvalidMoneyOperation {
        oneDollar.add(twoFrancs);
    }

    @Test(expected=InvalidMoneyOperation.class) public void cantSubtractDollarsAndFrancs() throws InvalidMoneyOperation {
        oneDollar.sub(twoFrancs);
    }

    @Test public void canApplyTheMoneyOnABank() {
        Bank bank = mock(Bank.class);
        oneDollar.applyOn(bank);
        verify(bank).apply(oneDollar);
    }
}
